src/otaclient/client_package.py

```diff
@@ -0,0 +1,590 @@
+                    continue
+
+                _mount_point = f"{mount_base}{_path}"
+                cmdhelper.ensure_mointpoint(
+                    _mount_point,
+                    ignore_error=False,
+                )
+                cmdhelper.ensure_mount(
+                    target=_path,
+                    mnt_point=_mount_point,
+                    mount_func=mount_func,
+                    raise_exception=True,
+                )
+
+        # bind necessary directories
+        # should not include /bin, /lib, /usr, /sbin because they are included in the squashfs
+        RW_PATHS = [
+            "/boot",
+            "/boot/firmware",
+            "/dev",
+            "/dev/shm",
+            "/etc",
+            "/home",
+            "/mnt",
+            "/opt",
+            "/ota-cache",
+            "/proc",
+            "/root",
+            "/run",
+            "/sys",
+            "/tmp",
+            "/usr/share/ca-certificates",
+            "/var",
+        ]
+
+        bind_paths(
+            paths=RW_PATHS, mount_base=mount_base, mount_func=cmdhelper.bind_mount_rw
+        )
+
+    def _rbind_mount_current_root(self, mount_base: StrOrPath) -> None:
+        """Mount the active slot to the mount base."""
+        # After chroot, the active slot root is not accessible from the chroot environment.
+        # So we need to bind mount the active slot before chroot.
+
+        # check if the mount base exists
+        if not os.path.exists(mount_base):
+            raise ValueError(f"Mount base does not exist: {mount_base}")
+
+        _mount_point = f"{mount_base}{cfg.DYNAMIC_CLIENT_MNT_ORIGINAL_ROOT}"
+        logger.info(f"mounting {cfg.ACTIVE_ROOT} to {_mount_point}")
+        cmdhelper.ensure_mointpoint(
+            _mount_point,
+            ignore_error=True,
+        )
+
+        cmdhelper.ensure_mount(
+            target=cfg.ACTIVE_ROOT,
+            mnt_point=_mount_point,
+            mount_func=cmdhelper.rbind_mount_ro,
+            raise_exception=True,
+        )
+
+    def _bind_mount_active_slot(self, mount_base: StrOrPath) -> None:
+        """Mount the active slot to the mount base."""
+        # After chroot, the active slot root is not accessible from the chroot environment.
+        # So we need to bind mount the active slot before chroot.
+
+        # check if the mount base exists
+        if not os.path.exists(mount_base):
+            raise ValueError(f"Mount base does not exist: {mount_base}")
+
+        _mount_point = f"{mount_base}{cfg.ACTIVE_SLOT_MNT}"
+        logger.info(f"mounting {cfg.ACTIVE_ROOT} to {_mount_point}")
+        cmdhelper.ensure_mointpoint(
+            _mount_point,
+            ignore_error=True,
+        )
+
+        cmdhelper.ensure_mount(
+            target=cfg.ACTIVE_ROOT,
+            mnt_point=_mount_point,
+            mount_func=cmdhelper.bind_mount_ro,
+            raise_exception=True,
+        )
```

Comment:
I know the purpose of this `_bind_mount_active_slot` function now, but it doesn't work as expected I think.

Here the target is `cfg.ACTIVE_ROOT`, which has value `/`. But for dynamic otaclient, `/` is the squashfs image, not the actual host root. You might need to change the `target` to `mount_base`.

Also, when mounting host rootfs, let's remove the `--make-unbindable` flag in `cmdhelper.rbind_mount_ro`, we don't need to worry about unexpected bind mount anymore as we are using `unshare`.